vm/src/slots.rs

```diff
@@ -148,10 +163,176 @@ pub trait SlotDescriptor: PyValue {
 pub trait Hashable: PyValue {
     #[pyslot]
     fn tp_hash(zelf: PyObjectRef, vm: &VirtualMachine) -> PyResult<PyHash> {
-        let zelf = <PyRef<Self>>::try_from_object(vm, zelf)?;
+        let zelf = PyRef::try_from_object(vm, zelf)?;
         Self::hash(zelf, vm)
     }
 
     #[pymethod(magic)]
     fn hash(zelf: PyRef<Self>, vm: &VirtualMachine) -> PyResult<PyHash>;
 }
+
+#[pyimpl]
+pub trait Comparable: PyValue {
+    #[pyslot]
+    fn tp_cmp(
+        zelf: PyObjectRef,
+        other: PyObjectRef,
+        op: PyComparisonOp,
+        vm: &VirtualMachine,
+    ) -> PyResult<PyComparisonValue> {
+        let zelf = PyRef::try_from_object(vm, zelf)?;
+        Self::cmp(zelf, other, op, vm)
+    }
+
+    fn cmp(
```

Comment:
Adding comment to idiomatically add `#[inline]` for impls of this trait will be helpful. We actually only need to inline `cmp`s which is called inside of RustPython code, but that kind of tracking is not easy. Adding `#[inline]` for every impl is easy and hopefully will not be inlined if it is too bad to inline - if llvm is good enough.

Reply:
Actually, I don't think it would optimize very much, cause the `__op__` methods will only be called when explicitly called like `x.__gt__(other)` from Python; `x > other` will still call `(cls.slots.cmp)(a, b, PyComparisonOp::Gt, vm)`

Reply:
ah, unfortune. maybe we can try multiple slot items for each cmp with proper benchmarks later